mountpoint-s3/src/prefetch/caching_stream.rs

```diff
@@ -0,0 +1,422 @@
+            }
+        };
+
+        pin_mut!(get_object_result);
+        let mut block_index = block_range.start;
+        let mut buffer = ChecksummedBytes::default();
+        loop {
+            match get_object_result.next().await {
+                Some(Ok((offset, body))) => {
+                    trace!(offset, length = body.len(), "received GetObject part");
+
+                    // Split the body into blocks.
+                    let mut body: Bytes = body.into();
+                    while !body.is_empty() {
+                        let remaining = (block_size as usize).saturating_sub(buffer.len()).min(body.len());
+                        let chunk = body.split_to(remaining);
+                        if let Err(e) = buffer.extend(chunk.into()) {
+                            error!(key, error=?e, "Integrity check failed");
+                            self.part_queue_producer.push(Err(e.into()));
+                            return;
+                        }
+                        if buffer.len() < block_size as usize {
+                            break;
+                        }
+                        self.update_cache(block_index, &buffer);
+                        self.part_queue_producer
+                            .push(Ok(self.make_part(buffer, block_index, &range)));
+                        block_index += 1;
+                        buffer = ChecksummedBytes::default();
+                    }
+                }
+                Some(Err(e)) => {
+                    error!(key, error=?e, "GetObject body part failed");
+                    self.part_queue_producer
+                        .push(Err(PrefetchReadError::GetRequestFailed(e)));
+                    break;
+                }
+                None => {
+                    if !buffer.is_empty() {
+                        if buffer.len() + (block_index * block_size) as usize == range.object_size() {
+                            // Write last block to the cache.
+                            self.update_cache(block_index, &buffer);
+                        }
+                        self.part_queue_producer
+                            .push(Ok(self.make_part(buffer, block_index, &range)));
+                    }
+                    break;
+                }
+            }
+        }
+        trace!("request finished");
+    }
+
+    fn update_cache(&self, block_index: u64, block: &ChecksummedBytes) {
+        // TODO: consider updating the cache asynchronously
+        let start = Instant::now();
+        match self.cache.put_block(self.cache_key.clone(), block_index, block.clone()) {
+            Ok(()) => {
+                metrics::histogram!("cache.write_duration_us", start.elapsed().as_micros() as f64);
+                metrics::counter!("cache.total_bytes", block.len() as u64, "type" => "write");
+            }
+            Err(error) => {
+                warn!(key=?self.cache_key.s3_key, block_index, ?error, "failed to update cache");
```

Comment:
Not for this PR: metric on cache failures? We should take a pass over all our cache metrics once we're happy with the implementation.